cmd/oras/internal/option/target.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package option
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/oci"
+	"oras.land/oras-go/v2/registry"
+)
+
+const (
+	TargetTypeRemote    = "remote"
+	TargetTypeOCILayout = "oci"
+)
+
+// Unary target option struct.
+type Target struct {
+	Remote
+	RawReference string
+	Type         string
+	TagOrDigest  string
+
+	isOCI bool
+}
+
+// ApplyFlags applies flags to a command flag set for unary target
+func (opts *Target) ApplyFlags(fs *pflag.FlagSet) {
+	opts.applyFlagsWithPrefix(fs, "", "")
+	opts.Remote.ApplyFlags(fs)
+}
+
+// AnnotatedReference returns full printable reference.
+func (opts *Target) AnnotatedReference() string {
+	return fmt.Sprintf("[%s] %s", opts.Type, opts.RawReference)
+}
+
+// applyFlagsWithPrefix applies flags to a command flag set with a prefix string.
+// Commonly used for non-unary remote targets.
+func (opts *Target) applyFlagsWithPrefix(fs *pflag.FlagSet, prefix, description string) {
+	var (
+		flagPrefix string
+		noteSuffix string
+	)
+	if prefix != "" {
+		flagPrefix = prefix + "-"
+		noteSuffix = description + " "
+	}
+	fs.BoolVarP(&opts.isOCI, flagPrefix+"oci", "", false, "Set "+noteSuffix+"target as an OCI-layout.")
+}
+
+// ApplyFlagsWithPrefix applies flags to a command flag set with a prefix string.
+// Commonly used for non-unary remote targets.
+func (opts *Target) ApplyFlagsWithPrefix(fs *pflag.FlagSet, prefix, description string) {
+	opts.applyFlagsWithPrefix(fs, prefix, description)
+	opts.Remote.ApplyFlagsWithPrefix(fs, prefix, description)
+}
+
+// Parse gets target options from user input.
+func (opts *Target) Parse() error {
+	switch {
+	case opts.isOCI:
+		opts.Type = TargetTypeOCILayout
+	default:
+		opts.Type = TargetTypeRemote
+	}
+	return nil
+}
+
```

Comment:
Kind of feels like there must be some general image uri parsing that breaks up a uri into registry, repository, tag, digest.

Reply:
The allowed characters of file paths and image references are different. For instance, registry and repository can only be lower letters.

@qweeah We may need to document the reference format like `<path>[:<tag>|@<digest>]`. In other words, the reference
```
C:\Test\repo:v1
```
is also valid although it contains two colons.

Here are more test cases:
- Windows path without a tag or digest: `C:\Test\repo`. A scenario could be `oras cp` to an OCI layout without a tag.
- Windows path with digest `C:\Test\repo@sha256:xxxx`.

Reply:
Yeh, that ParseReference is pretty limited since the Reference object doesn't have enough fields. Just seems like there should be an object that better represents these things.

Reply:
`C:\Test\repo` is indeed a corner case if we split the right-most colon. Reused `fileref.Parse` to cover that and unit test added.